service/discovery/aws/compute.go

```diff
@@ -0,0 +1,243 @@
+type LambdaAPI interface {
+	ListFunctions(ctx context.Context,
+		params *lambda.ListFunctionsInput, optFns ...func(*lambda.Options)) (*lambda.ListFunctionsOutput, error)
+}
+
+// newFromConfigEC2 holds ec2.NewFromConfig(...) allowing a test function to mock it
+var newFromConfigEC2 = ec2.NewFromConfig
+
+// newFromConfigLambda holds lambda.NewFromConfig(...) allowing a test function tp mock it
+var newFromConfigLambda = lambda.NewFromConfig
+
+// NewComputeDiscovery constructs a new awsS3Discovery initializing the s3-virtualMachineAPI and isDiscovering with true
+func NewComputeDiscovery(client *Client) discovery.Discoverer {
+	return &computeDiscovery{
+		virtualMachineAPI: newFromConfigEC2(client.Cfg),
+		functionAPI:       newFromConfigLambda(client.Cfg),
+		isDiscovering:     true,
+		awsConfig:         client,
+	}
+}
+
+// Name is the method implementation defined in the discovery.Discoverer interface
+func (d computeDiscovery) Name() string {
+	return "AWS Compute"
+}
+
+// List is the method implementation defined in the discovery.Discoverer interface
+func (d computeDiscovery) List() (resources []voc.IsCloudResource, err error) {
+	listOfVMs, err := d.discoverVirtualMachines()
+	if err != nil {
+		return
+	}
+	for _, machine := range listOfVMs {
+		resources = append(resources, &machine)
+	}
+
+	listOfFunctions, err := d.discoverFunctions()
+	if err != nil {
+		return
+	}
+	for _, function := range listOfFunctions {
+		resources = append(resources, &function)
+	}
+
+	return
+}
+
+// discoverVirtualMachines discovers all VMs (in the current region)
+func (d *computeDiscovery) discoverVirtualMachines() ([]voc.VirtualMachine, error) {
+	resp, err := d.virtualMachineAPI.DescribeInstances(context.TODO(), &ec2.DescribeInstancesInput{})
+	if err != nil {
+		var ae smithy.APIError
+		if errors.As(err, &ae) {
+			err = formatError(ae)
+		}
+		return nil, err
+	}
+	var resources []voc.VirtualMachine
+	for _, reservation := range resp.Reservations {
+		for i := range reservation.Instances {
+			vm := &reservation.Instances[i]
+			computeResource := &voc.Compute{
+				CloudResource: &voc.CloudResource{
+					ID:   d.getARNOfVM(vm),
+					Name: d.getNameOfVM(vm),
+					// TODO(all): -1 or 0 as default value when no creation time is available?
+					CreationTime: -1,
+					Type:         []string{"VirtualMachine", "Compute", "Resource"},
+				},
+			}
+
+			resources = append(resources, voc.VirtualMachine{
+				Compute:          computeResource,
+				NetworkInterface: d.getNetworkInterfacesOfVM(vm),
+				BlockStorage:     d.getBlockStorageIDsOfVM(vm),
+				Log:              d.getLogsOfVM(vm),
+			})
+		}
+	}
+	return resources, nil
+}
+
+// discoverFunctions discovers all lambda functions
+// TODO(all): FunctionVersion in input to ALL ok? (=all versions of a lambda functions are discovered)
+// TODO(oxisto): Is this a good approach with "for hasNextMarker" loop (I would use overloaded methods - not in Go)
```

Comment:
Not sure, it seems to be a little bit of copy/paste code. Can we combine the two ListFunctions calls some how?

Reply:
I found a possible solution: I loop from the beginning (`for true`) and in the first execution, the `Marker` in the input of `ListFunctions` is set to nil. At the end of the loop it is checked if the response has a `Marker` returned: If yes, execute `ListFunctions` again with the respective `Marker` and break, otherwise.